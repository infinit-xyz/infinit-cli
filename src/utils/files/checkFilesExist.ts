import { chalkError, chalkSuccess } from '@constants/chalk'
import type { FolderStructure } from '@utils/files/files.type'

import fs from 'fs'
import _ from 'lodash' // [TODO/INVESTIGATE] later on importing from lodash
import path from 'path'
import { match } from 'ts-pattern'

/**
 *
 * @param folderStructure Expected folder structure
 * @param folderPath Path to the folder
 * @param mode for 'all' all of the required files must exactly match the existing files , for 'require' only the required files must exist
 * @param isDisableLog Disable log
 * @returns {boolean} true if the files are equal, false otherwise
 */
export const checkFilesExist = (
  folderStructure: FolderStructure,
  folderPath: string,
  mode: 'all' | 'require' = 'all',
  isDisableLog: boolean = false,
): boolean => {
  const files = fs.readdirSync(folderPath)
  const expectedFiles = Object.keys(folderStructure)

  const sortedFiles = _.sortBy(files)
  const sortedExpectedFiles = _.sortBy(expectedFiles)
  const isFilesEqual = match(mode)
    .with('all', () => _.isEqual(sortedFiles, sortedExpectedFiles))
    .with('require', () => expectedFiles.every((file) => files.includes(file)))
    .exhaustive()

  if (!isFilesEqual) {
    if (!isDisableLog) {
      console.error(chalkError(`Files are not equal on path ${folderPath}`))

      console.error(`\nExpected: `)
      console.error(chalkSuccess(sortedExpectedFiles.join(', ')))
      console.error(`\nActual: `)
      console.error(chalkError(sortedFiles.join(', ')))
    }

    return false
  }

  for (const key in folderStructure) {
    const filePath = path.join(folderPath, key)
    if (folderStructure[key] === true) {
      if (!fs.existsSync(filePath)) {
        if (!isDisableLog) {
          console.error(chalkError(`File ${filePath} does not exist`))
        }
        return false
      }
    } else {
      const newFolderStructure = folderStructure[key] as FolderStructure
      if (!checkFilesExist(newFolderStructure, filePath, mode, isDisableLog)) {
        return false
      }
    }
  }

  return true
}
