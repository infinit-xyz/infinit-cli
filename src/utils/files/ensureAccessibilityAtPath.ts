import fs from 'fs'

/**
 * Check if the file exists and accessible
 * @param path file path
 * @param permissionMode fs.constants
 */
export const ensureAccessibilityAtPath = (path: string, permissionMode?: number) => {
  try {
    fs.accessSync(path, permissionMode ?? fs.constants.F_OK)
  } catch (_) {
    throw new Error('Permission required, run the command with sudo permission')
  }
}
