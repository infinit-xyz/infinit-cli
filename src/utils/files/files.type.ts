export interface FolderStructure {
  [index: string]: true | FolderStructure
}
