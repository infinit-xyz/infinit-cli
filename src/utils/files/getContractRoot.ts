import path from 'path'
import { config } from '@classes'

/**
 * Get contract root
 * @returns {string} contractRootPath
 */
export const getContractRoot = (): string => {
  const projectConfig = config.getProjectConfig()

  // replace _ with -
  const protocolModule = projectConfig.protocol_module.replaceAll('_', '-')

  const contractRootPath = path.resolve(process.cwd(), 'node_modules', `@infinit-xyz/${protocolModule}`, 'contracts')

  return contractRootPath
}
