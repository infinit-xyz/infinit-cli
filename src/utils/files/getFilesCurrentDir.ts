import fs from 'fs'

export const getFilesCurrentDir = (_path?: string): string[] => {
  const path = _path ?? process.cwd()
  if (!fs.existsSync(path)) {
    return []
  }

  const files = fs.readdirSync(path)
  return files
}
