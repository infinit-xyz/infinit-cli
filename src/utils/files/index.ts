export * from './checkFilesExist'
export * from './ensureAccessibilityAtPath'
export * from './ensureCwdRootProject'
export * from './getFilesCurrentDir'
export * from './isCwdRootProject'
export * from './readProjectRegistry'
export * from './writeFileSync'
