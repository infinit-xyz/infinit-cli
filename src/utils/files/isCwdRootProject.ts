import { FILE_NAMES } from '@constants'
import { checkFilesExist } from '@utils/files/checkFilesExist'

/**
 * Determines if the current working directory (CWD) is the root of the project.
 *
 * This function checks if the CLI is running from the root directory of the project
 * by comparing the CLI path with the current working directory and verifying the existence
 * of specific configuration files.
 *
 * @param {string} [currentCwd=process.cwd()] - The current working directory. Defaults to the process's current working directory.
 * @returns {Object} An object containing:
 * - `currentCwd` (string): The current working directory.
 * - `isRunningFromRootProject` (boolean): A flag indicating whether the CLI is running from the root project directory.
 */
export const isCwdRootProject = (currentCwd = process.cwd()) => {
  // Get the path of the CLI which depending on the package manager
  // npm, bun, yarn: */<project-name>/node_modules/.bin/infinit
  // pnpm: */<project-name>/node_modules/<cli-package-name>/bin/cli.js
  const cliPath = process.argv[1]

  const isCwdDirectoryMatch = cliPath.includes(currentCwd)

  if (!isCwdDirectoryMatch) {
    return { currentCwd, isRunningFromRootProject: false }
  }

  // */project-name
  const isRunningFromRootProject = checkFilesExist({ src: { [FILE_NAMES.CONFIG]: true } }, currentCwd, 'require', true)

  return { currentCwd, isRunningFromRootProject }
}
