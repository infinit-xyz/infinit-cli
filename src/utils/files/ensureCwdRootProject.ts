import chalk from 'chalk'

import { FILE_NAMES } from '@constants'
import { chalkError } from '@constants/chalk'
import { isCwdRootProject } from '@utils/files/isCwdRootProject'

/**
 * Ensure the current working directory is the root of the project
 * If not, exit the process with code 1
 */
export const ensureCwdRootProject = () => {
  const { currentCwd, isRunningFromRootProject } = isCwdRootProject()

  if (!isRunningFromRootProject) {
    console.error(chalkError(`Can't find ${chalk.bold(FILE_NAMES.CONFIG)}`))
    console.error(`Current path: ${currentCwd}`)
    process.exit(1)
  }
}
