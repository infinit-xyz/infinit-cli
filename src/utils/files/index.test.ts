import * as files from '@utils/files'
import { describe, expect, test } from 'vitest'

describe('@utils/files', () => {
  test('should match snapshot', () => {
    expect(files).toMatchSnapshot()
  })
})
