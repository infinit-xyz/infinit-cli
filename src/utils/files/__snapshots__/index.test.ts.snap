// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@utils/files > should match snapshot 1`] = `
{
  "checkFilesExist": [Function],
  "ensureAccessibilityAtPath": [Function],
  "ensureCwdRootProject": [Function],
  "getFilesCurrentDir": [Function],
  "isCwdRootProject": [Function],
  "readProjectRegistry": [Function],
  "writeFileSync": [Function],
}
`;
